### PR TITLE
Added a protected property to configure the properties that can be serialized within the Mailer/Message

### PR DESCRIPTION
**Pull request by josedaian**
With this new change it is possible to better extend the Mailer/Message class to be able to customize the properties that can be serializable.

For example:
```php
<?php
declare(strict_types=1);

namespace App\Mailer;

use Cake\Mailer\Message;

class CustomMessage extends Message {
    protected $serializableProperties = [
        'to', 'from', 'sender', 'replyTo', 'cc', 'bcc', 'subject',
        'returnPath', 'readReceipt', 'emailFormat', 'emailPattern', 'domain',
        'attachments', 'messageId', 'headers', 'appCharset', 'charset', 'headerCharset',
        'textMessage'
    ];
}
```

### CODE REVIEW

File: src/Mailer/Message.php [hunk 1]
```diff
@@ -301,6 +301,18 @@ class Message implements JsonSerializable, Serializable
      */
     protected $emailPattern = self::EMAIL_PATTERN;
 
+    /**
+     * Properties that could be serialized
+     *
+     * @var string[]
```
Comment (ADmad) [inline src/Mailer/Message.php]:
```suggestion
     * @var array<string>
```

**Comment by markstory** [comment]
> With this new change it is possible to better extend the Mailer/Message class to be able to customize the properties that can be serializable.

What problems does this help solve?

It would be good to have a test covering this behavior. We don't generally merge features without tests as they regress too easily down the road.

**Comment by othercorey** [comment]
@josedaian Can you help explain your use case?

**Comment by josedaian** [comment]
> @josedaian Can you help explain your use case?

I had a case where I wanted to save the serialized message, and there were properties that did not save me.
Going to the class and the function that generated the serialization, I found that the properties to serialize were only defined inside the jsonSerializable function.

So what I basically did was create a class that extends the Message class, and override the jsonSerializable function.

```php
class CustomMessage extends Message {
    protected $serializableProperties = [
        'to', 'from', 'sender', 'replyTo', 'cc', 'bcc', 'subject',
        'returnPath', 'readReceipt', 'emailFormat', 'emailPattern', 'domain',
        'attachments', 'messageId', 'headers', 'appCharset', 'charset', 'headerCharset',
        'textMessage', 'htmlMessage'
    ];

    public function jsonSerialize(): array
    {
        $array = [];
        foreach ($this->serializableProperties as $property) {
            $array[$property] = $this->{$property};
        }

        array_walk($array['attachments'], function (&$item, $key): void {
            if (!empty($item['file'])) {
                $item['data'] = $this->readFile($item['file']);
                unset($item['file']);
            }
        });

        return array_filter($array, function ($i) {
            return $i !== null && !is_array($i) && !is_bool($i) && strlen($i) || !empty($i);
        });
    }
}
```

So, long story short, if you cast that local property inside that function to a property of the class, it would be easier to extend without needing to override and repeat the function that serializes them.

**Review by ADmad — APPROVED** [review]

**Comment by markstory** [comment]
If you are dropping properties from the object when serializing, won't rebuilding the object be lossy?

**Comment by josedaian** [comment]
> If you are dropping properties from the object when serializing, won't rebuilding the object be lossy?

That is a possibility, but it would be nice if you could have that control and in this way only the properties that are necessary within the created Message class would be serialized.

That is, there may be properties that I am not interested in serializing, and I can intentionally configure it.

**Comment by markstory** [comment]
Ok. Any chance we can get a test for the limiting/filtering behavior of this new property? Without that we're likely to end up breaking this behavior in the future.

**Comment by markstory** [comment]
Thank you 🎉
